### PR TITLE
Add possibility to refresh tokens

### LINKED ISSUE #16: Add refresh tokens

### PR DESCRIPTION
**Pull request by FelixGail**
GET /token
- Now returns not only a token but also a refreshtoken

Get /refresh
- new endpoint
- Takes a refreshtoken and returns a new pair of tokens

close #16

### CODE REVIEW

File: MusicBot.yaml [hunk 1]
```diff
@@ -777,6 +793,20 @@ components:
...
+      required:
+        - token
+        - refreshToken
+      properties:
+        token:
+          description: A short-lived jwt token issued by the server.
+          type: string
+        refreshToken:
+          description: A long-livived jwt token to refresh the token.
+          type: string
```
Comment (FelixGail) [inline MusicBot.yaml]: Should properties be named `accessToken` and `refreshToken`?

Comment (BjoernPetersen) [reply MusicBot.yaml]: Yes

File: MusicBot.yaml [hunk 2]
```diff
@@ -148,6 +148,22 @@ paths:
...
+        pair of tokens
+      responses:
+        200:
+          description: An authorization token.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Tokens'
+        401:
+          $ref: '#/components/responses/Unauthenticated'
```
Comment (BjoernPetersen) [inline MusicBot.yaml]: 1. I'm not happy about the endpoint name. An endpoint should always be named after the resource you're accessing, not the action you're performing.
1. This operation uses the same security scheme as the other operations. I would introduce another security scheme specifically for the refresh token.
1. I think I would just merge this operation with the `loginUser` operation by adding another `securityScheme` to the latter.

File: MusicBot.yaml [hunk 3]
```diff
@@ -777,13 +778,32 @@ components:
...
+        A Token and a refresh token.
+      required:
+        - accessToken
+        - refreshToken
+      properties:
+        accessToken:
+          description: A short-lived jwt token issued by the server.
+          type: string
+        refreshToken:
+          description: A long-livived jwt token to refresh the token.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
          description: A long-lived jwt token to refresh the access token.
```

**Comment by BjoernPetersen** [comment]
I'm not sure about some details:

### Revocation

Refresh tokens should probably be revokable. How do we do that?
If we store all valid tokens somewhere anyway, we don't need to bother with JWT signing.

One idea I had on this was to store an individual refresh-token-signature-key for each user.

Pro:

- Can be efficiently cached, it's just a single key for each user
- All of a users refresh tokens can be revoked at once by regenerating the signature key

Con:

- Key generation is kind of expensive

### Key refresh

If we end up storing issued refresh tokens, we probably wouldn't want to create a new one every time a new access token is requested. Or do we?

### Lifetime

**How** long-lived should a refresh token be?

**Comment by FelixGail** [comment]
## Lifetime

I read some that the normal lifetime of a refresh token should be between 60 and 90 days. But since many users will only use this bot during parties, which in a normal household not seldomly have more than 3 month between them, i would increase the lifetime to half a year.

## Revocation

I haven't considered revocation.

### Alternative to signature

Instead of saving a signature, a jwt token could contain a random string as revocation claim. Simply deleting the string server-side invalidates the token.

### Granularity of revocation
If the key is revocable, it should be considered to save a key for each device connected to a user.
This would imply multiple new endpoints:
- GET /device
    - Retrieves a list of connected devices including an identifier for their refresh token
- GET /token/{tokenId}
    - Retrieve detailed information about a specific token
- DELETE /token/{tokenId}
    - Revokes a token
It also opens a discussion about when to remove outdated tokens. With a routine on startup?
If a new token is granted with a new login - should there be a maximum of concurrent tokens?

## Key refresh

If the key is not refreshed, how is the lifetime of a key implemented?
Just updating the lifetime makes tokens inexpireable, not updating the lifetime will make a user eventually logout, even with frequent use.

 -------

## Is it worth?

Does the whole jwt token scheme still make sense if this is implemented?
If tokens are revokeable, have long lifetime and are not refreshed. Wouldn't it be easier to just drop the jwt token completely and add an endpoint to retrieve user permissions, always authenticating with the above token?

**Comment by BjoernPetersen** [comment]
### Revocation

I think storing a string for every issued token defeats the purpose of JWT. A JWT token should be validated by signature without any database lookup. Storing the strings in memory moves the database lookup to the app initialization time, but we still need to write a string to database every few minutes and, as you already pointed out, have to delete outdated strings from time to time.

If a random revocation claim string is included in the JWT refresh token, we might as well just send that string as the whole token instead of using JWT.

I'm not a fan of this because we need to micro-manage keys/token/strings so much. A per-user signature key combined with the JWT expiration time claim would vastly simplify things.
The only drawback is the inflexibility for users, because they can't revoke a single refresh token. I don't think that will be a common occurrence though, so it should be fine.

### Refresh

I agree with the requirement for refresh token renewal, but I don't think it should happen as frequently as access token refresh. 

Suggestion:

- Make the `refreshToken` response property optional
- Only include a refresh token if the used refresh token will expire "soon"
- Add a query parameter flag (`includeRefresh`?) to enforce the inclusion of a new refresh token

### Lifetime 

I agree, 6 months seem appropriate.

File: MusicBot.yaml [hunk 4]
```diff
@@ -777,13 +798,31 @@ components:
         isSupported:
           description: Whether volume control is actually supported by the bot.
           type: boolean
+    Tokens:
+      type: object
+      description: >
+        A Token and optionally a refresh token.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
        An access token and optionally a refresh token.
```

File: MusicBot.yaml [hunk 5]
```diff
@@ -148,6 +156,19 @@ paths:
             text/plain:
               schema:
                 type: string
+    delete:
+      operationId: invalidateToken
+      summary: Invalidate all refresh-tokens for the current user.
+      description: >
+        Invalidates all refresh-tokens for the current user.
+        Access-tokens will still be valid until expiration.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
        Access tokens will still be valid until expiration.
```

File: MusicBot.yaml [hunk 6]
```diff
@@ -148,6 +156,19 @@ paths:
...
+      responses:
+        200:
+          description: successfully invalidated.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
          description: Successfully invalidated.
```

File: MusicBot.yaml [hunk 7]
```diff
@@ -126,18 +126,26 @@ paths:
...
         - Basic: []
+        - RefreshToken: []
       description: >
         Retrieves an Authorization token for a user.
+      parameters:
+        - name: includeRefresh
+          in: query
+          schema:
+            type: boolean
+          description: Enforce inclusion of a new refresh-token.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
          description: Enforce inclusion of a new refresh token.
```

File: MusicBot.yaml [hunk 8]
```diff
@@ -126,18 +126,26 @@ paths:
   /token:
     get:
       operationId: loginUser
-      summary: Retrieves a token for a user
+      summary: Retrieves a token and a refresh token for the user
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
      summary: Retrieves a token and optionally a refresh token for the user.
```

File: MusicBot.yaml [hunk 9]
```diff
@@ -126,18 +126,26 @@ paths:
...
       security:
         - Basic: []
+        - RefreshToken: []
       description: >
         Retrieves an Authorization token for a user.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
        Retrieves an access token for a user. The response may also also include a refresh token, 
        if Basic authentication or a soon-to-expire refresh token was used in the request.
```

File: MusicBot.yaml [hunk 10]
```diff
@@ -777,13 +798,31 @@ components:
...
           type: boolean
+    Tokens:
+      type: object
+      description: >
+        A Token and optionally a refresh token.
+      required:
+        - accessToken
+      properties:
+        accessToken:
+          description: A short-lived jwt token issued by the server.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
          description: >
            A short-lived JWT token issued by the server.
            This token is used to authenticate all further API requests, except token refreshes.
```

File: MusicBot.yaml [hunk 11]
```diff
@@ -777,13 +798,31 @@ components:
...
+          type: string
+        refreshToken:
+          description: A long-lived jwt token to refresh the access token.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
          description: A long-lived opaque token to refresh the access token.
```

File: MusicBot.yaml [hunk 12]
```diff
@@ -777,13 +798,31 @@ components:
...
     Token:
       type: http
       scheme: bearer
       description: An API key issued by server.
       bearerFormat: jwt
+    RefreshToken:
+      type: http
+      scheme: bearer
+      description: A jwt refresh-token issued by the server.
+      bearerFormat: jwt
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
      description: A refresh token issued by the server.
      bearerFormat: opaque
```

File: MusicBot.yaml [hunk 13]
```diff
@@ -148,6 +157,19 @@ paths:
             text/plain:
               schema:
                 type: string
+    delete:
+      operationId: invalidateToken
+      summary: Invalidate all refresh-tokens for the current user.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
      summary: Invalidate all refresh tokens for the current user.
```

File: MusicBot.yaml [hunk 14]
```diff
@@ -148,6 +157,19 @@ paths:
...
+      description: >
+        Invalidates all refresh-tokens for the current user.
```
Comment (BjoernPetersen) [inline MusicBot.yaml]:
```suggestion
        Invalidates all refresh tokens for the current user.
```

**Pull request merged**